applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.h

```diff
@@ -0,0 +1,525 @@
+    ///@{
+
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    virtual void InitializeComputationOfMappingMatrix()
+    {
+        mpSearchTree.reset();
+        mMappingMatrix.clear();
+    }
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+
+    ///@}
+
+private:
+    ///@name Static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // Variables for spatial search
+    unsigned int mBucketSize = 100;
+    KDTree::Pointer mpSearchTree;
+
+    // Variables for mapping
+    SparseMatrixType mMappingMatrix;
+
+    SymmetryBase::Pointer mpSymmetry;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    // --------------------------------------------------------------------------
+    void CreateFilterFunction()
+    {
+        const std::string filter_type = mMapperSettings["filter_function_type"].GetString();
+        const double filter_radius = mMapperSettings["filter_radius"].GetDouble();
+
+        mpFilterFunction = Kratos::shared_ptr<FilterFunction>(new FilterFunction(filter_type, filter_radius));
+    }
+
+    // --------------------------------------------------------------------------
+    void InitializeMappingVariables()
+    {
+        const unsigned int origin_node_number = mrOriginModelPart.Nodes().size();
+        const unsigned int destination_node_number = mrDestinationModelPart.Nodes().size();
+        mMappingMatrix.resize(destination_node_number*3,origin_node_number*3,false);
+    }
+
+    // --------------------------------------------------------------------------
+    void AssignMappingIds()
+    {
+        unsigned int i = 0;
+        for(auto& node_i : mrOriginModelPart.Nodes())
+            node_i.SetValue(MAPPING_ID,i++);
+
+        i = 0;
+        for(auto& node_i : mrDestinationModelPart.Nodes())
+            node_i.SetValue(MAPPING_ID,i++);
+    }
+
+    // --------------------------------------------------------------------------
+    void CreateSearchTreeWithAllNodesInOriginModelPart()
+    {
+        mpSearchTree = Kratos::shared_ptr<KDTree>(new KDTree(mpSymmetry->GetOriginSearchNodes().begin(), mpSymmetry->GetOriginSearchNodes().end(), mBucketSize));
+    }
+
+    // --------------------------------------------------------------------------
+    void ComputeMappingMatrix()
+    {
+        const double filter_radius = mMapperSettings["filter_radius"].GetDouble();
+        const unsigned int max_number_of_neighbors = mMapperSettings["max_nodes_in_filter_radius"].GetInt();
+
+        // variable size, fixed capacity vecs
+        std::vector<bool> transform;
+        NodeVector total_neighbor_nodes;
+        std::vector<double> total_list_of_weights;
+        std::vector<double> list_of_weights;
+        transform.reserve(max_number_of_neighbors);
+        total_neighbor_nodes.reserve( max_number_of_neighbors );
+        total_list_of_weights.reserve( max_number_of_neighbors );
+        list_of_weights.reserve( max_number_of_neighbors );
```

Comment:
```suggestion
        std::vector<bool> transform(max_number_of_neighbors);
        NodeVector total_neighbor_nodes;
        std::vector<double> total_list_of_weights;
        std::vector<double> list_of_weights;
        total_neighbor_nodes.reserve( max_number_of_neighbors );
        total_list_of_weights.reserve( max_number_of_neighbors );
        list_of_weights.reserve( max_number_of_neighbors );
```

same for others

Reply:
Why should i not use reserve?

Reply:
you should:
~~~c++
std::vector<bool> transform;
transform.reserve(max_number_of_neighbors);
~~~
||
v
~~~c++
std::vector<bool> transform(max_number_of_neighbors);
~~~

Reply:
Yeah, this i got. But its not the same, since my version creates a vector with size 0 but a capacity of max_number_of_neighbors, while yours creates a vector of size max_number_of_neighbors.

But i see it makes the code easier to read, and in the end its the same in terms of memory.

Reply:
Edit: it actually makes the code more complicated (IMO) because i need to keep track of the amount of data that has already been filled into the vector instead of simply inserting to the end.

Reply:
sorry I misread, I thought it was `resize` :man_facepalming: 

sorry for the noise